scheds/rust/scx_rustland/src/bpf/main.bpf.c

```diff
@@ -513,8 +517,22 @@ void BPF_STRUCT_OPS(rustland_dispatch, s32 cpu, struct task_struct *prev)
 		dbg_msg("usersched: pid=%d cpu=%d payload=%llu",
 			task.pid, task.cpu, task.payload);
 		dispatch_task(p, task.cpu, 0);
-		__sync_fetch_and_add(&nr_user_dispatches, 1);
 		bpf_task_release(p);
+		__sync_fetch_and_add(&nr_user_dispatches, 1);
+
+		/*
+		 * Wake-up the target CPU if it's different than the current
+		 * one. This allows the task to run immediately, even if the
+		 * target CPU is not awake.
+		 */
+		if (task.cpu != cpu)
+			scx_bpf_kick_cpu(task.cpu, 0);
+	}
+
+	/* Consume all tasks enqueued in the current CPU's DSQ */
+	bpf_repeat(MAX_ENQUEUED_TASKS) {
+		if (!scx_bpf_consume(cpu))
```

Comment:
Let's say a task is enqueued in usersched, and then has its affinity updated, and then usersched dispatches the task targeting a CPU which is no longer in the task's cpumask. It wouldn't fail right away but wouldn't the task end up in the wrong dsq and never be consumed?

Reply:
> Let's say a task is enqueued in usersched, and then has its affinity updated, and then usersched dispatches the task targeting a CPU which is no longer in the task's cpumask. It wouldn't fail right away but wouldn't the task end up in the wrong dsq and never be consumed?

The usersched doesn't have any knowledge of the cpumask, it assumes that all tasks can run on all CPUs.

It's the BPF dispatcher that validates, in `.dispatch()`, if the target CPU is allowed or not, according to the cpumask. If it's not allowed, it will attempt to dispatch the task to the prev CPU, if that is also not allowed we pick a random CPU from the allowed ones, according to the cpumask.

So, if the affinity changes while the task is enqueued in the usersched, no problem, the BPF part would still bounce the task to a valid CPU.

However, if the affinity changes while the `.dispatch()` is executed (but I'm not sure if that's possible, I assume there's some locking involved while the `.dispatch()` callback is executed), then there could be a race and we may end up consuming the task on a CPU that is not usable by the task.

So the question is, can cpumask change concurrently while `.dispatch()` is executed?

Reply:
`set_cpus_allowed()` is exclusive to scheduling operations through `.dequeue()`. rustland doesn't implement `.dequeue()` but the intended design is that sched_ext core still keeps track of task state so that it can automatically ignore spurious dispatches which take place after dequeue (and followup re-enqueue). Hmm.. I'm not quite sure why this didn't work.

If dispatch and cpumask update are racing, what matters is 1. whether `finish_dispatch()` claims the task before the task gets dequeued by `set_cpus_allowed()` path, and if so 2. whether the task gets put on the target CPU's local dsq before dequeue. If either fails, the dispatch attempt is ignored as they can be treated as the task getting first dispatched and then dequeued. If both succeeded, the task is in local dsq and then dequeued from it.

You're hitting the `scx_ops_error()` call at the end of `dispatch_to_local_dsq()` without this patch, right? Hmm... right, the rq locked here can be a different one from the one the target task is associated with, but `DISPATCHING` is set, which should block `ops_dequeue()`.

I reproduce the problem and understand why it ends up dispatching to the wrong CPU. What should I do to reproduce?

Reply:
@htejun it's very easy, try `stress-ng --race-sched 8 --timeout 10`, it should trigger the ops error almost immediately:
```
09:18:51 [WARN] FAIL: SCX_DSQ_LOCAL[_ON] verdict target cpu 3 not allowed for stress-ng-race-[1004999] (err=1024)
```